sdk/go/pulumi/context.go

```diff
@@ -1204,6 +1202,10 @@ func (state *resourceState) resolve(ctx *Context, err error, inputs *resourceInp
 		}
 	}
 
+	// We need to wait until after we finish mutating outprops to resolve. Resolving
+	// unlocks multithreaded access to the resolved value, making mutation a data race.
+	state.rawOutputs.getState().resolve(outprops, true, false, nil)
```

Comment:
This makes sense to me at some basic level. resolve(outprops) shares outprops with other goroutines, but it has not finished populating in the original code. So doing it after `outprops["id"]=`.. stanzas seems to be better. The `outprops[""]` I'd love to understand what that does and what for.

Reply:
It looks like a marker indicating either started to resolve (`resource.NewObjectProperty(remaining)`), finished resolving (`resource.NewObjectProperty(remaining)` where `remaining = resource.PropertyMap{}`) or will not resolve (`resource.MakeComputed(resource.NewStringProperty(""))`).

I'm not 100% sure though.

Reply:
Grepping for the key `""` in `sdk/go/pulumi`, its only uses are in `resolve` function. Despite that, remove it results in test failures for `TestRegisterResource`.